Added User requests to display posts and comments and random user ID

simonrowlands opened the pull request:
**Users Screen**
After tapping the "Get posts and Comments" button, a zipped network request should perform the following:
- Network request should be made to get posts and the returned post count displayed in posts label
- Network request should be made to get comments and the returned post count displayed in posts label

After tapping the "Get User and their Post count" button, a network request should be made to:
- Fetch Users from the /users endpoint, select a random User and update User ID label
- Fetch Posts from /posts endpoint using the userID from the previous request and update Post Count label with this users post count

simonrowlands committed:
Adding new models

simonrowlands committed:
Subclassing NetworkingAPI

simonrowlands committed:
Adding UsersViewController + ViewModel
Downloads posts/comments in Rx.zip and displays in label

simonrowlands committed:
User requests to display random user and post count

tailec commented on NCU-Project/API/BaseNetworkingAPI.swift:
```diff
@@ -0,0 +1,29 @@
+
+final class BaseURL {
+    
+    var baseURL: String
```
can make it `private let`

tailec commented on NCU-Project/API/UsersNetworkingAPI.swift:
```diff
@@ -0,0 +1,69 @@
+                users.randomElement()
+            }
+    }
+}
```
nice 💯

simonrowlands committed:
Extracting ViewModel out of the ViewController

simonrowlands committed:
Adding network activity indicator

tailec commented on NCU-Project/Users/UsersViewController.swift:
```diff
@@ -0,0 +1,60 @@
+        
+        output.userPostCount.bind { [weak self] postCount in
+            self?.userPostsCountLabel.text = "Post Count: " + String(postCount)
+        }.disposed(by: disposeBag)
```
I think you could do `output.userPostCount.bind(to: .userPostsCountLabel.rx.text).disposed(by: disposeBag)`

and in the view model do:
commentsCount: getPostsResponse.map { return $0.0.count }.map { "Post Count: " + String($0) }` 

^ same would apply for other labels

tailec commented on NCU-Project/Users/UsersViewModel.swift:
```diff
@@ -0,0 +1,58 @@
+final class UsersViewModel: ViewModelType {
+    
+    struct Input {
+        let postsButtonTap: ControlEvent<Void>
```
instead of `ControlEvent`, pass `Observable` and in view controller do 
`let input = UsersViewModel.Input(postsButtonTap: getPostsButton.rx.tap.asObservable(), userButtonTap: getUserButton.rx.tap.asObservable())`

tailec commented on NCU-Project/Users/UsersViewModel.swift:
```diff
@@ -0,0 +1,58 @@
+        
+        let activityIndicator = ActivityIndicator()
+        
+        let getPostsResponse = input.postsButtonTap.flatMap {
```
`flatMapLatest`

tailec commented on NCU-Project/Users/UsersViewModel.swift:
```diff
@@ -0,0 +1,58 @@
+                .trackActivity(activityIndicator)
+                .observeOn(MainScheduler.instance)
+        }.share()
+        
```
let's improve formatting a little bit and swap `observeOn` like:
```
let getUserResponse = input.userButtonTap
  .flatMapLatest {
            UsersNetworkingAPI.getRandomUser()
                .trackActivity(activityIndicator)
  }
  .observeOn(MainScheduler.instance)
  .share()
```

simonrowlands committed:
Changing view model output to contain strings
Directly binding labels to newly returned strings
Changed viewModel input to use observable rather than control event
Changed .zip to flatMapLatest to be more efficient

simonrowlands committed:
Fixed network request.. should have been passing in URLQuery instead of URLParams

simonrowlands committed:
Some layout and moving static function to perform on its instance

wrumble commented on NCU-Project/API/DogsNetworkingAPI.swift:
```diff
@@ -0,0 +1,37 @@
+    
+    static let baseURL = BaseURL("https://dog.ceo/api")
+    
+    static let allDogsList = "/breeds/list/all"
```
is it worth extracting the path strings you have dotted around into a enum/struct so they are all in one place?

wrumble commented on NCU-Project/API/UsersNetworkingAPI.swift:
```diff
@@ -0,0 +1,68 @@
+    
+    static var baseURL = BaseURL("https://jsonplaceholder.typicode.com")
+    
+    static var posts = "/posts"
```
these could be added to the Path struct/enum

wrumble commented on NCU-Project/Model/Users/Post.swift:
```diff
@@ -0,0 +1,39 @@
+
+extension Post: Decodable {
+    
+    enum PostKeys: String, CodingKey {
```
could this be renamed to just Keys so that it is called Post.Keys instead of Post.PostKeys?

wrumble commented on NCU-Project/Model/Users/User.swift:
```diff
@@ -0,0 +1,38 @@
+
+extension User: Decodable {
+    
+    enum UserKeys: String, CodingKey {
```
same here

simonrowlands committed:
Cleaning up the project
Adding associated type to BaseNetworkingAPI
Adding enums to NetworkingAPI's

simonrowlands committed:
Extracting API enums into an Endpoint struct
Renaming BaseNetworkingAPI -> Networkable to be more protocol-y
